File: ascii_art.java
// ascii_art.java
import java.awt.*;
import java.awt.image.*;
import java.io.*;
import java.nio.file.*;
import java.util.*;
import javax.imageio.*;

public class ascii_art {
    private static final String RESET = "\u001B[0m";
    private static final String BOLD = "\u001B[1m";

    private static String rgbToAnsi(int r, int g, int b) {
        return String.format("\u001B[38;2;%d;%d;%dm", r, g, b);
    }

    private static final String[][] CHARSETS = {
        {" ", ".", ":", "-", "=", "+", "*", "#", "%", "@"},           // default
        {" ", ".", "o", "O", "0", "@"},                               // simple
        {" ", "`", ".", ",", ":", ";", "+", "*", "?", "%", "S", "#", "@"}, // detailed
        {" ", "░", "▒", "▓", "█"}                                     // block
    };

    private static int getTerminalWidth() {
        try {
            return Integer.parseInt(System.getenv("COLUMNS"));
        } catch (Exception e) {
            return 80;
        }
    }

    private static String imageToAscii(String path, int width, int height, int charsetIdx, boolean invert, boolean color) throws Exception {
        BufferedImage img = ImageIO.read(new File(path));
        if (img == null) throw new Exception("Unsupported format");
        int origW = img.getWidth(), origH = img.getHeight();

        if (width == 0 && height == 0) {
            width = getTerminalWidth() / 2;
        }
        if (width == 0) width = origW * height / origH;
        if (height == 0) height = origH * width / origW;
        if (width < 1) width = 1;
        if (height < 1) height = 1;

        Image scaled = img.getScaledInstance(width, height, Image.SCALE_SMOOTH);
        BufferedImage resized = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
        Graphics2D g = resized.createGraphics();
        g.drawImage(scaled, 0, 0, null);
        g.dispose();

        String[] chars = CHARSETS[charsetIdx];
        StringBuilder sb = new StringBuilder();
        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                int rgb = resized.getRGB(x, y);
                int r = (rgb >> 16) & 0xFF;
                int g2 = (rgb >> 8) & 0xFF;
                int b = rgb & 0xFF;
                int gray = (int)(0.299 * r + 0.587 * g2 + 0.114 * b);
                if (invert) gray = 255 - gray;
                int idx = gray * (chars.length - 1) / 255;
                if (color) sb.append(rgbToAnsi(r, g2, b));
                sb.append(chars[idx]);
            }
            if (color) sb.append(RESET);
            sb.append('\n');
        }
        return sb.toString();
    }

    private static void batchConvert(String inputDir, String outputDir, int width, int height, int charsetIdx, boolean invert, boolean color, boolean verbose) throws Exception {
        Files.createDirectories(Paths.get(outputDir));
        String[] exts = {".png", ".jpg", ".jpeg", ".bmp", ".tiff", ".webp"};
        Set<String> extSet = new HashSet<>(Arrays.asList(exts));
        File[] files = new File(inputDir).listFiles();
        if (files == null) return;
        int total = 0;
        java.util.List<File> imageFiles = new ArrayList<>();
        for (File f : files) {
            if (f.isFile() && extSet.contains(f.getName().substring(f.getName().lastIndexOf('.')).toLowerCase())) {
                imageFiles.add(f);
            }
        }
        total = imageFiles.size();
        for (int i = 0; i < total; i++) {
            File f = imageFiles.get(i);
            String outPath = outputDir + File.separator + f.getName().replaceFirst("\\.[^.]+$", "") + ".txt";
            if (verbose) {
                System.out.println("[" + (i+1) + "/" + total + "] " + f.getName() + " -> " + outPath);
            }
            String ascii = imageToAscii(f.getPath(), width, height, charsetIdx, invert, color);
            Files.write(Paths.get(outPath), ascii.getBytes());
            if (!verbose) {
                int pct = (i+1) * 100 / total;
                System.out.printf("\r[%s] %d%% %d/%d", "█".repeat(pct/2) + "░".repeat(50-pct/2), pct, i+1, total);
            }
        }
        if (!verbose) System.out.println();
        System.out.println("✅ Batch conversion completed. Results in " + outputDir);
    }

    public static void main(String[] args) throws Exception {
        String input = null, output = null;
        int width = 0, height = 0, charsetIdx = 0;
        boolean invert = false, noColor = false, batch = false, verbose = false;

        for (int i = 0; i < args.length; i++) {
            if (args[i].equals("-o") && i+1 < args.length) output = args[++i];
            else if (args[i].equals("-w") && i+1 < args.length) width = Integer.parseInt(args[++i]);
            else if (args[i].equals("-h") && i+1 < args.length) height = Integer.parseInt(args[++i]);
            else if (args[i].equals("-c") && i+1 < args.length) {
                String cs = args[++i];
                if (cs.equals("default")) charsetIdx = 0;
                else if (cs.equals("simple")) charsetIdx = 1;
                else if (cs.equals("detailed")) charsetIdx = 2;
                else if (cs.equals("block")) charsetIdx = 3;
            } else if (args[i].equals("-i")) invert = true;
            else if (args[i].equals("-n")) noColor = true;
            else if (args[i].equals("-b")) batch = true;
            else if (args[i].equals("-v")) verbose = true;
            else if (args[i].equals("-h") || args[i].equals("--help")) {
                System.out.println("Usage: ascii_art <input> [options]\n  -o <file>   Output file/dir\n  -w <N>      Width\n  -h <N>      Height\n  -c <set>    Charset\n  -i          Invert\n  -n          No color\n  -b          Batch\n  -v          Verbose");
                return;
            } else if (input == null) input = args[i];
        }
        if (input == null) { System.err.println("Укажите входной файл или папку."); System.exit(1); }

        if (batch) {
            if (output == null) { System.err.println("Укажите выходную папку для batch."); System.exit(1); }
            batchConvert(input, output, width, height, charsetIdx, invert, !noColor, verbose);
        } else {
            String ascii = imageToAscii(input, width, height, charsetIdx, invert, !noColor);
            if (output != null) {
                Files.write(Paths.get(output), ascii.getBytes());
                System.out.println("✅ Result saved to " + output);
            } else {
                System.out.print(ascii);
            }
        }
    }
}

File: ascii_art.cs
// ascii_art.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.Processing;
using SixLabors.ImageSharp.PixelFormats;

class AsciiArt
{
    static string Colorize(string text, string color) => color + text + "\x1b[0m";
    static string RgbToAnsi(byte r, byte g, byte b) => $"\x1b[38;2;{r};{g};{b}m";

    static Dictionary<string, char[]> CHARSETS = new Dictionary<string, char[]>
    {
        {"default", new char[]{' ','.',':','-','=','+','*','#','%','@'}},
        {"simple", new char[]{' ','.','o','O','0','@'}},
        {"detailed", new char[]{' ','`','.',',',':',';','+','*','?','%','S','#','@'}},
        {"block", new char[]{' ','░','▒','▓','█'}}
    };

    static int GetTerminalWidth()
    {
        try { return Console.WindowWidth; } catch { return 80; }
    }

    static string ImageToAscii(string path, int width, int height, string charsetName, bool invert, bool color)
    {
        using var img = Image.Load<Rgb24>(path);
        int origW = img.Width, origH = img.Height;

        if (width == 0 && height == 0)
            width = GetTerminalWidth() / 2;
        if (width == 0)
            width = origW * height / origH;
        if (height == 0)
            height = origH * width / origW;
        if (width < 1) width = 1;
        if (height < 1) height = 1;

        img.Mutate(ctx => ctx.Resize(width, height));

        var chars = CHARSETS.ContainsKey(charsetName) ? CHARSETS[charsetName] : CHARSETS["default"];
        var sb = new StringBuilder();
        for (int y = 0; y < img.Height; y++)
        {
            for (int x = 0; x < img.Width; x++)
            {
                var pixel = img[x, y];
                int gray = (int)(0.299 * pixel.R + 0.587 * pixel.G + 0.114 * pixel.B);
                if (invert) gray = 255 - gray;
                int idx = gray * (chars.Length - 1) / 255;
                if (color)
                    sb.Append(RgbToAnsi(pixel.R, pixel.G, pixel.B));
                sb.Append(chars[idx]);
            }
            if (color) sb.Append("\x1b[0m");
            sb.Append('\n');
        }
        return sb.ToString();
    }

    static void BatchConvert(string inputDir, string outputDir, int width, int height, string charset, bool invert, bool color, bool verbose)
    {
        Directory.CreateDirectory(outputDir);
        var exts = new HashSet<string>{".png",".jpg",".jpeg",".bmp",".tiff",".webp"};
        var files = Directory.GetFiles(inputDir, "*.*", SearchOption.TopDirectoryOnly)
            .Where(f => exts.Contains(Path.GetExtension(f).ToLower()))
            .ToList();
        int total = files.Count;
        for (int i = 0; i < total; i++)
        {
            string outPath = Path.Combine(outputDir, Path.GetFileNameWithoutExtension(files[i]) + ".txt");
            if (verbose)
                Console.WriteLine($"[{i+1}/{total}] {files[i]} -> {outPath}");
            string ascii = ImageToAscii(files[i], width, height, charset, invert, color);
            File.WriteAllText(outPath, ascii);
            if (!verbose)
            {
                int pct = (i+1) * 100 / total;
                Console.Write($"\r[{new string('█', pct/2)}{new string('░', 50-pct/2)}] {pct}% {i+1}/{total}");
            }
        }
        if (!verbose) Console.WriteLine();
        Console.WriteLine($"✅ Batch conversion completed. Results in {outputDir}");
    }

    static int Main(string[] args)
    {
        string input = null, output = null;
        int width = 0, height = 0;
        string charset = "default";
        bool invert = false, noColor = false, batch = false, verbose = false;

        for (int i = 0; i < args.Length; i++)
        {
            if (args[i] == "-o" && i+1 < args.Length) output = args[++i];
            else if (args[i] == "-w" && i+1 < args.Length) width = int.Parse(args[++i]);
            else if (args[i] == "-h" && i+1 < args.Length) height = int.Parse(args[++i]);
            else if (args[i] == "-c" && i+1 < args.Length) charset = args[++i];
            else if (args[i] == "-i") invert = true;
            else if (args[i] == "-n") noColor = true;
            else if (args[i] == "-b") batch = true;
            else if (args[i] == "-v") verbose = true;
            else if (args[i] == "-h" || args[i] == "--help")
            {
                Console.WriteLine("Usage: ascii_art <input> [options]\n  -o <file>   Output file/dir\n  -w <N>      Width\n  -h <N>      Height\n  -c <set>    Charset: default, simple, detailed, block\n  -i          Invert\n  -n          No color\n  -b          Batch\n  -v          Verbose");
                return 0;
            }
            else if (input == null) input = args[i];
        }
        if (input == null) { Console.Error.WriteLine("Укажите входной файл или папку."); return 1; }

        if (batch)
        {
            if (output == null) { Console.Error.WriteLine("Укажите выходную папку для batch."); return 1; }
            BatchConvert(input, output, width, height, charset, invert, !noColor, verbose);
        }
        else
        {
            string ascii = ImageToAscii(input, width, height, charset, invert, !noColor);
            if (output != null)
            {
                File.WriteAllText(output, ascii);
                Console.WriteLine($"✅ Result saved to {output}");
            }
            else
            {
                Console.Write(ascii);
            }
        }
        return 0;
    }
}
